ref: #4807 4.1.x用にcanonical url追加

tao-s opened the pull request:
## 概要(Overview・Refs Issue)
ref: #4807

## マイナーバージョン互換性保持のための制限事項チェックリスト

- [ ] 既存機能の仕様変更
- [x] フックポイントの呼び出しタイミングの変更
- [x] フックポイントのパラメータの削除・データ型の変更
- [x] twigファイルに渡しているパラメータの削除・データ型の変更
- [x] Serviceクラスの公開関数の、引数の削除・データ型の変更
- [x] 入出力ファイル(CSVなど)のフォーマット変更

## レビュワー確認項目

- [x] 動作確認
- [x] コードレビュー
- [x] E2E/Unit テスト確認(テストの追加・変更が必要かどうか)
- [x] 互換性が保持されているか
- [x] セキュリティ上の問題がないか

tao-s committed:
ref: #4807 canonical url追加

okazy commented:
@tao-s 
すみません、反応が遅くなりました。
こちら間違ってクローズされたのではないでしょうか？

動作確認しました。
おおよそ動作問題ないかと思います。

検索ページで `Category` が null で定義されているのでエラーとなってしまっていました。
それぞれで null のチェックも必要かと思います。

```twig
    {# canonical url #}
    {% if Category is defined and Category is not null %}
    <link rel="canonical" href="{{url('product_list')}}?category_id={{Category.id}}" />
    {% elseif Product is defined and Product is not null %}
    <link rel="canonical" href="{{url('product_detail',{"id":Product.id})}}" />
    {% elseif Page is defined and Page is not null %}
    <link rel="canonical" href="{{url(Page.url)}}" />
    {% endif %}
```

一覧ページのページネーションについて、こちらの実装ではカテゴリページの全てのページに同じ canonical が設定されます。
Google の推奨設定では、各ページでコンテンツ（商品）が異なるため別ページとして扱うのが良いようです。

> Googleウェブマスター向け公式ブログより
> 間違い 1: 複数ページにまたがるコンテンツの 1 ページ目を rel=canonical のリンク先とする
> https://webmaster-ja.googleblog.com/2013/05/5-common-mistakes-with-relcanonical.html

Google の推奨設定のためには、全商品表示の商品ページを作成するか、rel="canonical" の代わりに rel=”prev” と rel=”next” を利用する必要がありますが、実装が冗長になってしまう懸念があります。

私の意見としては、推奨設定ではありませんが、ひとまずはいただいている実装で問題ないのではと思いました。

tao-s commented:
@okazy こちら解決したんで復活させます

tao-s committed:
商品一覧のページング、user_data配下にユーザが作成したページに対応

tao-s commented:
c6d4c6a9f8c0c1efe1054c0930e6f914168c34d8 でcanonicalがmeta.twigに移されたので、商品一覧のページング処理などそっちにもってく

nanasess committed:
Merge remote-tracking branch 'upstream/4.1' into seo-canonical

nanasess committed:
Fix canonical url

- Fix  `Unexpected character "&".`
- Category.title は存在しないため Category.name に修正
- ページング等の処理を meta.twig へ移動

nanasess commented:
@tao-s 最新の 4.1 ブランチに追従して、修正を以下に PR しましたのでご確認いただけますでしょうか？
https://github.com/tao-s/ec-cube-1/pull/7

matsuoshi commented:
ありがとうございます、
@nanasess さんの修正を @tao-s さんのこのPRにマージいたしました

matsuoshi commented:
マイページの注文履歴詳細のところでテストが落ちているので確認中です

- マイページ注文履歴詳細 URL: `/mypage/history/{order_no}`
- 商品購入/お届け先の追加 URL: `/shopping/shipping_edit/{id}`

など、URLにパラメタが必須のページについて、canonical URLを設定しようとしてエラーになっている模様
https://github.com/EC-CUBE/ec-cube/pull/4809/files#diff-ebd2f65646a5a6ff6780c9e04fbfe24f7725aff31b8113ec82b41041305c8722R25

Commit:
システムページについて canonical を作成しないよう変更

matsuoshi commented:
システムのデフォルトのページについて、先のコメントのように URLにパラメタが必須のページの場合、エラーとなっておりました。今回は

- 商品一覧・詳細など、主要なページのcanonicalは設定できている
- システムのデフォルトページについてはcanonicalをつける優先度は低い

と判断して、システムデフォルトページの場合 canonical 生成をスキップするように修正し、4.1.2 に取り込めればと考えています。
@tao-s いかがでしょうか？

chihiro-adachi committed:
日本語をメッセージファイルに切り出し

chihiro-adachi committed:
cs fix

chihiro-adachi commented:
`○ページ目`の記述をメッセージファイルに切り出しました。
その他動作確認しましたが問題ないかと思います。

chihiro-adachi commented:
@tao-s 
いったんマージさせていただきました。
こちら一部挙動を変更していますので、もう少しこうしたい、などありましたらPRいただければと。
https://github.com/EC-CUBE/ec-cube/pull/4809#issuecomment-1021937166

ご対応ありがとうございました。
